src/app.rs

```diff
@@ -0,0 +1,266 @@
+        };
+
+        let config = {
+            let config_paths = root_opts.config_paths.clone();
+            let watch_config = root_opts.watch_config;
+            let require_healthy = root_opts.require_healthy;
+
+            rt.block_on(async move {
+                if let Some(s) = sub_command {
+                    let code = match s {
+                        SubCommand::Validate(v) => validate::validate(&v, color).await,
+                        SubCommand::List(l) => list::cmd(&l),
+                        SubCommand::Test(t) => unit_test::cmd(&t),
+                        SubCommand::Generate(g) => generate::cmd(&g),
+                        #[cfg(windows)]
+                        SubCommand::Service(s) => service::cmd(&s),
+                    };
+
+                    return Err(code);
+                };
+
+                let config_paths = config::process_paths(&config_paths).ok_or(exitcode::CONFIG)?;
+
+                if watch_config {
+                    // Start listening for config changes immediately.
+                    config::watcher::spawn_thread(&config_paths, None).or_else(|error| {
+                        error!(message = "Unable to start config watcher.", %error);
+                        Err(exitcode::CONFIG)
+                    })?;
+                }
+
+                info!(
+                    message = "Loading configs.",
+                    path = ?config_paths
+                );
+
+                let config =
+                    config::load_from_paths(&config_paths).map_err(handle_config_errors)?;
+
+                config::LOG_SCHEMA
+                    .set(config.global.log_schema.clone())
+                    .expect("Couldn't set schema");
+
+                let diff = config::ConfigDiff::initial(&config);
+                let pieces = topology::build_or_log_errors(&config, &diff)
+                    .await
+                    .ok_or(exitcode::CONFIG)?;
+
+                #[cfg(feature = "api")]
+                let api = config.api;
+
+                let result = topology::start_validated(config, diff, pieces, require_healthy).await;
+                let (topology, graceful_crash) = result.ok_or(exitcode::CONFIG)?;
+
+                Ok(ApplicationConfig {
+                    config_paths,
+                    topology,
+                    graceful_crash,
+                    #[cfg(feature = "api")]
+                    api,
+                })
+            })
+        }?;
+
+        Ok(Application {
+            opts: root_opts,
+            config,
+            runtime: rt,
+        })
+    }
+
+    pub fn run(self) {
+        let mut rt = self.runtime;
+
+        let graceful_crash = self.config.graceful_crash;
+        let mut topology = self.config.topology;
+
+        let config_paths = self.config.config_paths;
+
+        #[cfg(feature = "api")]
+        // assigned to prevent the API terminating when falling out of scope
```

Comment:
(Not a change request) A rust note for you, if you called `_api` as `_` instead, this wouldn't work since Rust would automatically drop it. "Load bearing `_`"